src/warc2zim/utils.py

```diff
@@ -42,14 +53,53 @@ def parse_title(content):
         return ""
 
 
-def to_string(input_: str | bytes) -> str:
-    try:
-        input_ = input_.decode(  # pyright: ignore[reportGeneralTypeIssues, reportAttributeAccessIssue]
-            "utf-8-sig"
-        )
-    except AttributeError:
-        pass
-    return input_  # pyright: ignore[reportGeneralTypeIssues, reportReturnType]
+def get_record_encoding(record: ArcWarcRecord) -> str | None:
+    content_type = get_record_content_type(record)
+    if m := ENCODING_RE.search(content_type):
+        return m.group("encoding")
+
+
+def to_string(input_: str | bytes, encoding: str | None) -> str:
+    tried_encodings = set()
+    if isinstance(input_, str):
+        return input_
+
+    if not input_:
+        # Empty bytes are easy to decode
+        return ""
+
+    if encoding:
+        try:
+            return input_.decode(encoding)
+        except (ValueError, LookupError):
+            tried_encodings.add(encoding)
+            pass
+
+    # Detect encoding from content.
+    # First try declared encoding in content.
+    content_start = input_[:1024].decode("ascii", errors="replace")
+    if m := ENCODING_RE.search(content_start):
+        encoding = m.group("encoding")
+        if encoding and encoding not in tried_encodings:
+            try:
+                return input_.decode(encoding)
+            except (ValueError, LookupError):
+                tried_encodings.add(encoding)
+                pass
+
+    all_encodings = chardet.detect_all(input_)
+    encodings = [e["encoding"] for e in all_encodings if e["encoding"]]
+
+    if not encodings:
+        raise ValueError(f"Impossible to detect encoding of content {input_[:200]}")
```

Comment:
Either cover this line or add a pragma to indicate this is not feasible

Reply:
Removed. I never succeed to have this condition, and it is "covered" by the loop just after.